ironfish/src/memPool/recentlyEvictedCache.ts

```diff
@@ -0,0 +1,225 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import { BufferMap } from 'buffer-map'
+import { Assert } from '../assert'
+import { createRootLogger, Logger } from '../logger'
+import { MetricsMonitor } from '../metrics'
+import { getTransactionSize } from '../network/utils/serializers'
+import { Transaction, TransactionHash } from '../primitives/transaction'
+import { PriorityQueue } from './priorityQueue'
+
+interface EvictionEntry {
+  hash: TransactionHash
+  feeRate: bigint
+}
+
+interface ExpirationEntry {
+  hash: TransactionHash
+  expiration: number
+}
+
+export class RecentlyEvictedCache {
+  private readonly metrics: MetricsMonitor
+  private readonly logger?: Logger
+
+  private readonly capacity: number = 1000 // TODO: fix this default initialization
+
+  private readonly transactions = new BufferMap<Transaction>()
+  /**
+   * Keep track of number of bytes stored in the transaction map
+   * */
+  private transactionsBytes = 0
+
+  private readonly maxJailTime: number = 10 // TODO: fix -> this is max duration txns exist in cache
+
+  private currentSequence = 0
+
+  /**
+   * A priority queue of transactions to evict from this cache when full.
+   *
+   * Transactions are evicted in order of `feeRate` descending because
+   * they are most likely to be re-included into the mempool
+   */
+  private readonly evictionQueue: PriorityQueue<EvictionEntry>
+
+  private readonly expirationQueue: PriorityQueue<ExpirationEntry>
+
+  /**
+   * Creates a new cache to track transactions that have recently been evicted.
+   * Transactions are routinely evicted from the cache after the `maxJailTime`.
+   * When full, transactions are evicted in order of `feeRate` descending to
+   * make room for the new transaction.
+   * @param options
+   */
+  constructor(options: { metrics: MetricsMonitor; logger?: Logger }) {
+    const logger = options.logger || createRootLogger()
+
+    this.metrics = options.metrics
+    this.logger = logger.withTag('RecentlyEvictedCache')
+
+    this.evictionQueue = new PriorityQueue<EvictionEntry>(
+      (t1, t2) => t1.feeRate > t2.feeRate,
+      (t) => t.hash.toString('hex'),
+    )
+
+    this.expirationQueue = new PriorityQueue<ExpirationEntry>(
+      (t1, t2) => t1.expiration < t2.expiration,
+      (t) => t.hash.toString('hex'),
+    )
+  }
+
+  private remove(hash: TransactionHash): Transaction | void {
+    const transactionToRemove = this.get(hash)
+
+    if (!transactionToRemove) {
+      return
+    }
+
+    this.transactions.delete(hash)
```

Comment:
I think I would feel more comfortable keeping all these strongly in sync and removing from all queues at the same time. I'm trying to think of scenarios where a transaction gets removed by one process then re-added and it gets added to a queue twice somehow

Reply:
Also if we keep them in sync then we don't even need the `this.transactions` map we can just use one of the priority queues which has an internal map